design-documents/jwt-support.md

```diff
@@ -0,0 +1,180 @@
+        $this->deploymentConfig = $deploymentConfig;
+    }
+
+    public function create(): JWK
+    {
+        $secret = (string) $this->deploymentConfig->get('crypt/key');
+        return JWKFactory::createFromSecret(
+            $secret,
+            ['alg' => $this->algorithmFactory->getAlgorithmName(), 'use' => 'sig']
+        );
+    }
+}
+```
+
+The `\Magento\Framework\Jwt\GeneratorInterface` might be used to generate JWT based on claims:
+```php
+interface GeneratorInterface
+{
+    public function generate(array $claims = []): string;
+}
+
+class JwsGenerator implements GeneratorInterface
+{
+    public function __construct(
+        AlgorithmFactory $algorithmFactory,
+        CryptKeyGenerator $keyGenerator,
+        SerializerInterface $serializer
+    ) {
+        $this->algorithmFactory = $algorithmFactory;
+        $this->keyGenerator = $keyGenerator;
+        $this->serializer = $serializer;
+    }
+
+    public function generate(array $claims = []): string
+    {
+        $timestamp = time();
+        $baseClaims = [
+            'iat' => $timestamp,
+            'nbf' => $timestamp,
+            'exp' => $timestamp + 36000,
+            'use' => 'sig'
+        ];
+
+        $payload = json_encode(array_merge($claims, $baseClaims));
+
+        $jwsBuilder = new JWSBuilder(null, $this->algorithmFactory->getAlgorithmManager());
+        $jws = $jwsBuilder->create()
+            ->withPayload($payload)
+            ->addSignature($this->keyGenerator->create(), ['alg' => $this->algorithmFactory->getAlgorithmName()])
+            ->build();
+
+        return $this->serializer->serialize($jws);
+    }
+}
+```
+It will have implementations for JWS and JWE (the default preference will be for JWS implementation).
+
+The `\Magento\Framework\Jwt\VerifierInterface` provides a possibility to validate received JWT. The implementation might use claims validation for more advanced payload verification.
+```php
+interface VerifierInterface
+{
+    public function validate(string $token): bool;
+}
+
+class JwsVerifier implements VerifierInterface
+{
+    public function __construct(
+        AlgorithmFactory $algorithmFactory,
+        KeyGeneratorInterface $keyGenerator,
+        SerializerInterface $serializer,
+        ClaimCheckerManagerFactory $checkerManagerFactory
+    ) {
+        $this->algorithmFactory = $algorithmFactory;
+        $this->keyGenerator = $keyGenerator;
+        $this->serializer = $serializer;
+        $this->checkerManagerFactory = $checkerManagerFactory;
+    }
+
+    public function validate(string $token): bool
+    {
+        $verifier = $this->getVerifier();
+        $jws = $this->serializer->unserialize($token);
+
+        if (!$verifier->verifyWithKey($jws, $this->keyGenerator->create(), 0)) {
+            return false;
+        };
+
+        $checkers = ['iat', 'nbf', 'exp'];
+        $claimChecker = $this->checkerManagerFactory->add('iat', new IssuedAtChecker())
```

Comment:
there seem to be no way for 3rd-party developers to use Magento SPI to modify claims

Reply:
They are checkers, not claims. The checkers also should be configurable, it just an example from POC.